homeassistant/components/apple_tv/__init__.py

```diff
@@ -1,273 +1,289 @@
-                        vol.Required(CONF_LOGIN_ID): cv.string,
-                        vol.Optional(CONF_CREDENTIALS): cv.string,
-                        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
-                        vol.Optional(CONF_START_OFF, default=False): cv.boolean,
-                    }
-                )
-            ],
+    for domain in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, domain)
         )
-    },
-    extra=vol.ALLOW_EXTRA,
-)
 
-# Currently no attributes but it might change later
-APPLE_TV_SCAN_SCHEMA = vol.Schema({})
+    return True
 
-APPLE_TV_AUTHENTICATE_SCHEMA = vol.Schema({ATTR_ENTITY_ID: cv.entity_ids})
 
+async def async_unload_entry(hass, entry):
+    """Unload an Apple TV config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+                for platform in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        manager = hass.data[DOMAIN].pop(entry.unique_id)
+        await manager.disconnect()
+
+    return unload_ok
 
-def request_configuration(hass, config, atv, credentials):
-    """Request configuration steps from the user."""
-    configurator = hass.components.configurator
 
-    async def configuration_callback(callback_data):
-        """Handle the submitted configuration."""
+class AppleTVManager:
+    """Connection and power manager for an Apple TV.
 
-        pin = callback_data.get("pin")
+    An instance is used per device to share the same power state between
+    several platforms. It also manages scanning and connection establishment
+    in case of problems.
+    """
 
+    def __init__(self, hass, config_entry):
+        """Initialize power manager."""
+        self.config_entry = config_entry
+        self.hass = hass
+        self.listeners = []
+        self.message = None
+        self.atv = None
+        self._is_on = not config_entry.options.get(CONF_START_OFF, False)
+        self._connection_attempts = 0
+        self._connection_was_lost = False
+        self._task = None
+
+    async def init(self):
+        """Initialize power management."""
+        if self._is_on:
+            await self.connect()
+
+    def connection_lost(self, _):
+        """Device was unexpectedly disconnected."""
+        _LOGGER.warning('Connection lost to Apple TV "%s"', self.atv.name)
+        if self.atv:
+            self.atv.close()
+            self.atv = None
+        self._connection_was_lost = True
+        self._update_state(disconnected=True)
+        self._start_connect_loop()
+
+    def connection_closed(self):
+        """Device connection was (intentionally) closed."""
+        if self.atv:
+            self.atv.close()
+            self.atv = None
+        self._update_state(disconnected=True)
+        self._start_connect_loop()
```

Comment:
Yeah, we always want to maintain a connection in order to work as the user expects. If the Apple TV is turned off outside of home Assistant, this callback will be called (as it's an expected disconnect in that sense) and we shall try to set up the connection again. Otherwise the integration will be dead until it is manually turned on again.

Reply:
Ok.